cpp/include/graph.hpp

```diff
@@ -0,0 +1,213 @@
+
+namespace cugraph {
+namespace experimental {
+
+typedef enum prop_type{PROP_UNDEF, PROP_FALSE, PROP_TRUE} PropType;
+
+struct GraphProperties {
+  bool directed;
+  bool weighted;
+  bool multigraph;
+  bool bipartite;
+  bool tree;
+  PropType has_negative_edges;
+  GraphProperties() : directed(false), weighted(false), multigraph(false), bipartite(false), tree(false), has_negative_edges(PROP_UNDEF){}
+};
+
+/**
+ * @Synopsis    Base class graphs, all but vertices and edges
+ *
+ * @tparam VT   Type of vertex (defaults to int)
+ * @tparam WT   Type of weight (defaults to float)
+ */
+template <typename WT>
+class GraphBase {
+public:
+  WT const *edge_data;     ///< edge weight
+
+  GraphProperties          prop;
+
+  size_t                   number_of_vertices;
+  size_t                   number_of_edges;
+
+  GraphBase(WT const *edge_data_, size_t number_of_vertices_, size_t number_of_edges_):
+    edge_data(edge_data_),
+    prop(),
+    number_of_vertices(number_of_vertices_),
+    number_of_edges(number_of_edges_)
+  {}
+};
+
+/**
+ * @Synopsis    A graph stored in COO (COOrdinate) format.
+ *
+ * @tparam VT   Type of vertex (defaults to int)
+ * @tparam WT   Type of weight (defaults to float)
+ */
+template <typename VT = int, typename WT = float>
+class GraphCOO: public GraphBase<WT> {
+public:
+  VT const *src_indices;   ///< rowInd
+  VT const *dst_indices;   ///< colInd
+
+  /**
+   * @Synopsis   Default constructor
+   */
+  GraphCOO(): GraphBase<WT>(nullptr, 0, 0), src_indices(nullptr), dst_indices(nullptr) {}
+  
+  /**
+   * @Synopsis   Wrap existing arrays representing an edge list in a Graph.
+   *             GraphCOO does not own the memory used to represent this graph. This
+   *             function does not allocate memory.
+   *
+   * @Param  source_indices        This array of size E (number of edges) contains the index of the source for each edge.
+   *                               Indices must be in the range [0, V-1].
+   * @Param  destination_indices   This array of size E (number of edges) contains the index of the destination for each edge.
+   *                               Indices must be in the range [0, V-1].
+   * @Param  edge_data             This array size E (number of edges) contains the weight for each edge.  This array can be null
+   *                               in which case the graph is considered unweighted.
+   * @Param  number_of_vertices    The number of vertices in the graph
+   * @Param  number_of_edges       The number of edges in the graph
+   */
+  GraphCOO(VT const *src_indices_, VT const *dst_indices_, WT const *edge_data_,
+           size_t number_of_vertices_, size_t number_of_edges_):
+    GraphBase<WT>(edge_data_, number_of_vertices_, number_of_edges_),
+    src_indices(src_indices_),
+    dst_indices(dst_indices_)
+  {}
+};
+
+/**
+ * @Synopsis    Base class for graph stored in CSR (Compressed Sparse Row) format.
+ *
+ * @tparam VT   Type of vertex (defaults to int)
```

Comment:
I think we need ET as well (useful if # vertices < std::numeric_limits<int32_t>::max() but # edges > std::numeric_limits<int32_t>::max()). And specifying the default value, int32_t might be more specific than int (though these two are same in most systems).

Reply:
And shouldn't we better go for a safer option (int64_t & double) for default parameters as long as we're providing an option to override default parameters to save memory?

Reply:
Honestly, I'm not sure we should even provide a default.  I think the caller should explicitly identify what class they want.  I think I'll remove the defaults.

There's no legacy code using these graph objects, so there's nothing to gain by introducing this potential ambiguity.

Reply:
> Honestly, I'm not sure we should even provide a default
=>
I 100% agree. Better be explicit in this regard.